{"cells":[{"cell_type":"markdown","id":"00274e26","metadata":{},"source":["## The following are the codes demonstrating the performance of SERM imputation for scRNA-seq data"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# To illustrate the imputation calculation of SERM, the cellular taxonomy dataset analysed in the \n","# manuscript will be automatically loaded. However, you can upload your own dataset, and impute the \n","# data by SERM. Here the input data is saved as .mat file to reduce the data size (normally .csv file \n","# needs more disk space). However, .csv files can also be loaded using python packages such as \"pandas\""]},{"cell_type":"code","execution_count":4,"id":"f8e72a7a","metadata":{"collapsed":true},"outputs":[],"source":["# Import all the necessary Python packages\n","import os\n","from functools import partial\n","import h5py\n","import matplotlib.pyplot as plt\n","import scipy\n","import scipy.io as sio\n","from serm import *\n","from sklearn.decomposition import PCA\n","import numpy as np\n","from sklearn.manifold import TSNE\n","import magic\n","import random\n","%matplotlib inline"]},{"cell_type":"markdown","metadata":{},"source":["Load data"]},{"cell_type":"code","execution_count":5,"id":"aeab0cba","metadata":{},"outputs":[],"source":["# Load the original data (reference data), the downsampled data, and the data labels\n","# The original data is from the work of Baryawno et al. \n","# Reference: Baryawno, Ninib, et al. \"A cellular taxonomy \n","# of the bone marrow stroma in homeostasis and leukemia.\" Cell 177.7 (2019): 1915-1932.\n","# The data from the authors is very large with 23092 cells and 27998 genes. We choose 5% of the\n","# top highly variable genes (1093 genes) to create a smaller dataset.\n","# The following two code lines read the reference data described above\n","a=sio.loadmat('data/data_CellTax_Original.mat')\n","dataRef=a['dataRef']\n","\n","# The downsampled data (to 1%) were generated by using the technique described in Miao et al.\n","# Please see the description of their codes at \n","# https://www.bioconductor.org/packages/release/bioc/html/scRecover.html\n","# We used their \"countsSampling\" function for the downsampling operation\n","# Reference: Miao, Z., Li, J., & Zhang, X. (2019). scRecover: Discriminating true and \n","# false zeros in single-cell RNA-seq data for imputation. bioRxiv, 665323.\n","# The following two code lines read the downsampled data\n","a=sio.loadmat('data/data_CellTax_Downsampled_1perc.mat')\n","dataSample=a['dataSample']\n","\n","# Load data labels \n","a=sio.loadmat('data/dataLabel_CellTax.mat')\n","dataLabel=a['GT']"]},{"cell_type":"markdown","metadata":{},"source":["Set parameters of SERM and run it on the downsampled data"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Selection of the size of the region of interest (ROI)\n","# We choose ROI size in such a way that the expression matrix\n","# is divided into 8 (i.e. 4 by 2) ROIs. However, other ROI sizes are also\n","# fine as SERM is generally very robust against the ROI size. \n","ROIsize = [6000,600]\n","# Selection of percent of overlap between successive ROIs. The default value is 50%, but other \n","# values can be chosen\n","percOL = 0.5\n","# Run SERM\n","out_SERM = serm(dataSample,ROIsize,percOL,randomize=False) # If you want to randomize the rows of the data\n","# before applying SERM, please choose randomize=True"]},{"cell_type":"markdown","metadata":{},"source":["For performance comparison, run MAGIC with default parameters"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["magic_operator = magic.MAGIC()\n","out_MAGIC = magic_operator.fit_transform(dataSample)"]},{"cell_type":"markdown","metadata":{},"source":["Computation of Pearson coefficient for quantitative comparison"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Here, we compute the Pearson coefficient between the expression values of each gene of the cells from \n","# the reference data and imputed data. Thus, for the loaded dataset with 1093 genes, we get 1093 \n","# coefficients. We then use violin plot to show the  distribution of the coefficients\n","sz=dataSample.shape\n","corrVecDownsampled=np.zeros(sz[1])\n","corrVecSERM=np.zeros(sz[1])\n","corrVecMAGIC=np.zeros(sz[1])\n","\n","for i in range(sz[1]):\n","    \n","    downSampledGene=dataSample[:,i]\n","    referenceGene=dataRef[:,i]\n","    sermGene=out_SERM[:,i]\n","    \n","    a = np.corrcoef(downSampledGene, referenceGene)\n","    corrVecDownsampled[i]=a[1,0]\n","    \n","    b= np.corrcoef(sermGene, referenceGene)\n","    corrVecSERM[i]=b[1,0]\n","    \n","    magicGene=out_MAGIC[:,i]\n","    b= np.corrcoef(magicGene, referenceGene)\n","    corrVecMAGIC[i]=b[1,0]\n","    \n","# Stack all the coefficient values for plotting\n","corrVec=np.stack((corrVecDownsampled,corrVecMAGIC,corrVecSERM), axis=0)"]},{"cell_type":"markdown","metadata":{},"source":["Create violin plot of the Pearson coefficient values "]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["fig, (ax) = plt.subplots(nrows=1, ncols=1, sharey=True)\n","labels = ['Downsampled', 'MAGIC', 'SERM']\n","plt.violinplot(corrVec.T,showmeans=True)\n","plt.title('Comaprison of different imputation techniques')\n","set_axis_style(ax, labels)\n","ax.set_ylabel('Pearson coefficient')"]},{"cell_type":"markdown","metadata":{},"source":["t-SNE visualization of the reference, the downsampled, and the imputed data"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["tsne_Original = TSNE(n_components=2, learning_rate='auto',init='random',random_state=1).fit_transform(dataRef)\n","plt.figure(1)\n","plt.scatter(tsne_Original[:,0],tsne_Original[:,1], c=dataLabel.flatten(),s=1)\n","plt.xlabel(\"tSNE1\")\n","plt.ylabel(\"tSNE2\")\n","plt.title('t-SNE visualization of reference data')\n","\n","tsne_Downsampled = TSNE(n_components=2, learning_rate='auto',init='random',random_state=1).fit_transform(dataSample)\n","plt.figure(2)\n","plt.scatter(tsne_Downsampled[:,0],tsne_Downsampled[:,1], c=dataLabel.flatten(),s=1)\n","plt.xlabel(\"tSNE1\")\n","plt.ylabel(\"tSNE2\")\n","plt.title('t-SNE visualization of downsampled data')\n","plt.ylim(-30, 30)\n","\n","tsne_SERM = TSNE(n_components=2, learning_rate='auto',init='random',random_state=1).fit_transform(out_SERM)\n","plt.figure(3)\n","plt.scatter(tsne_SERM[:,0],tsne_SERM[:,1], c=dataLabel.flatten(),s=1)\n","plt.xlabel(\"tSNE1\")\n","plt.ylabel(\"tSNE2\")\n","plt.title('t-SNE visualization of SERM-imputed data')\n","\n","tsne_MAGIC = TSNE(n_components=2, learning_rate='auto',init='random',random_state=1).fit_transform(out_MAGIC)\n","plt.figure(4)\n","plt.scatter(tsne_MAGIC[:,0],tsne_MAGIC[:,1], c=dataLabel.flatten(),s=1)\n","plt.xlabel(\"tSNE1\")\n","plt.ylabel(\"tSNE2\")\n","plt.title('t-SNE visualization of MAGIC-imputed data')"]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.8"}},"nbformat":4,"nbformat_minor":5}
